rust/src/enhancers/grammar.rs

```diff
@@ -8,172 +8,266 @@
+use super::RegexCache;
+
+const MATCHER_LOOKAHEAD: [&str; 11] = [
+    "!",
+    "a",
+    "category:",
+    "e",
+    "f",
+    "me",
+    "mo",
+    "p",
+    "s",
+    "t",
+    "va",
+];
+
+fn expect<'a>(input: &'a str, pat: &str) -> anyhow::Result<&'a str> {
+    input
+        .strip_prefix(pat)
+        .ok_or_else(|| anyhow!("at `{input}`: expected `{pat}`"))
 }
 
-fn rule_bool(input: &str) -> IResult<&str, bool> {
-    alt((
-        value(true, alt((tag("1"), tag("yes"), tag("true")))),
-        value(false, alt((tag("0"), tag("no"), tag("false")))),
-    ))(input)
+fn bool(input: &str) -> anyhow::Result<bool> {
+    match input {
+        "1" | "yes" | "true" => Ok(true),
+        "0" | "no" | "false" => Ok(false),
+        _ => anyhow::bail!("at `{input}`: invalid boolean value"),
+    }
 }
 
-fn rule_number(input: &str) -> IResult<&str, usize> {
-    map(take_while1(|c: char| c.is_ascii_digit()), |n: &str| {
-        n.parse().unwrap()
-    })(input)
+fn ident(input: &str) -> anyhow::Result<(&str, &str)> {
+    let Some(end) =
+        input.find(|c: char| !(c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | '-')))
+    else {
+        return Ok((input, ""));
+    };
+
+    if end == 0 {
+        anyhow::bail!("at `{input}`: invalid identifier");
+    }
+
+    Ok(input.split_at(end))
 }
 
-fn frame_matcher(frame_offset: FrameOffset) -> impl Fn(&str) -> IResult<&str, Matcher> {
-    move |input| {
-        let input = input.trim_start();
-
-        let quoted_ident = delimited(
-            char('"'),
-            take_while1(|c: char| {
-                c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | ':' | '-' | ' ')
-            }),
-            char('"'),
-        );
-        let matcher_type = alt((ident, quoted_ident));
-
-        let unquoted = take_while1(|c: char| !c.is_ascii_whitespace());
-        // TODO: escapes, etc
-        let quoted = delimited(char('"'), take_while1(|c: char| c != '"'), char('"'));
-        let argument = alt((quoted, unquoted));
-
-        let mut matcher = map_res(
-            tuple((opt(char('!')), matcher_type, char(':'), argument)),
-            |(negated, matcher_type, _, argument): (_, _, _, &str)| {
-                // TODO: support even more escapes
-                let unescaped = argument.replace("\\\\", "\\");
-                Matcher::new(
-                    negated.is_some(),
-                    matcher_type,
-                    &unescaped,
-                    frame_offset,
-                    &mut Cache::default(),
-                )
-            },
-        );
-
-        matcher(input)
+fn argument(input: &str) -> anyhow::Result<(&str, &str)> {
+    if let Some(rest) = input.strip_prefix('"') {
+        let end = rest
+            .find('"')
+            .ok_or_else(|| anyhow!("at `{input}`: unclosed `\"`"))?;
+        let result = &rest[..end];
+        let rest = rest.get(end + 1..).unwrap_or_default();
```

Comment:
No need for `get` and `unwrap_or_else`. The `find` before has already asserted that we do have a trailing `"` which we can skip over.